sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,332 @@
+			}
+		}
+		tt.Fatal(driverErr)
+	}
+
+	defer func() {
+		if c.SkipTeardown {
+			logger.Info("driver Teardown skipped")
+			return
+		}
+		if err := c.Environment.Teardown(); err != nil {
+			logger.Error("error in driver teardown:", "error", err)
+		}
+	}()
+
+	// retrieve the root client from the Environment. If this returns an error, fail
+	// immediately
+	rootClient, err := c.Environment.Client()
+	if err != nil {
+		tt.Fatal(err)
+	}
+
+	// Trap the rootToken so that we can preform revocation or other tasks in the
+	// event any steps remove the token during testing.
+	rootToken := c.Environment.RootToken()
+
+	// Defer revocation of any secrets created. We intentionally enclose the
+	// responses slice so in the event of a fatal error during test evaluation, we
+	// are still able to revoke any leases/secrets created
+	var responses []*api.Secret
+	defer func() {
+		// restore root token for admin tasks
+		rootClient.SetToken(rootToken)
+		// failedRevokes tracks any errors we get when attempting to revoke a lease
+		// to log to users at the end of the test.
+		var failedRevokes []*api.Secret
+		for _, secret := range responses {
+			if secret.LeaseID == "" {
+				continue
+			}
+
+			if err := rootClient.Sys().Revoke(secret.LeaseID); err != nil {
+				tt.Error(fmt.Errorf("error revoking lease: %w", err))
+				failedRevokes = append(failedRevokes, secret)
+				continue
+			}
+		}
+
+		// If we have any failed revokes, log it.
+		if len(failedRevokes) > 0 {
+			for _, s := range failedRevokes {
+				tt.Error(fmt.Sprintf(
+					"WARNING: Revoking the following secret failed. It may\n"+
+						"still exist. Please verify:\n\n%#v",
+					s))
+			}
+		}
+	}()
+
+	stepCount := len(c.Steps)
+	for i, step := range c.Steps {
+		// range is zero based, so add 1 for a human friendly output of steps.
+		// "index" here is only used for logging / output, and not to reference the
+		// step in the slice of steps.
+		index := i + 1
+		if logger.IsWarn() {
+			progress := fmt.Sprintf("%d/%d", index, stepCount)
+			logger.Warn("Executing test step", "step_number", progress)
+		}
+
+		// reset token in case it was cleared
+		client, err := rootClient.Clone()
+		if err != nil {
+			tt.Fatal(err)
+		}
+
+		// TODO: support creating tokens with policies listed in each Step
+		client.SetToken(rootToken)
+
+		resp, respErr := makeRequest(tt, c.Environment, step)
+
+		if resp != nil {
+			responses = append(responses, resp)
+		}
+
+		// run the associated AssertionFunc, if any. If an error was expected it is
+		// sent to the Assert function to validate.
+		if step.Assert != nil {
+			if err := step.Assert(resp, respErr); err != nil {
+				tt.Error(fmt.Errorf("failed step %d: %w", index, err))
+			}
+		}
+	}
+}
+
+func makeRequest(tt TestT, env Environment, step Step) (*api.Secret, error) {
+	tt.Helper()
+	client, err := env.Client()
+	if err != nil {
+		return nil, err
+	}
+
+	if step.Unauthenticated {
+		client.ClearToken()
```

Comment:
Since `ClearToken` is updates the client's token value and the client is cached, we might want to defer adding back the client's original token after `makeRequest`.

Reply:
Great idea, I added that in 09b2307bd

Reply:
Great idea, I added that in 09b2307bd